actors/builtin/market/market_state.go

```diff
@@ -47,72 +50,36 @@ func ConstructState(emptyArrayCid, emptyMapCid, emptyMSetCid cid.Cid) *State {
 		EscrowTable:    emptyMapCid,
 		LockedTable:    emptyMapCid,
 		NextID:         abi.DealID(0),
-		DealIDsByParty: emptyMSetCid,
+		DealOpsByEpoch: emptyMSetCid,
 	}
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 // Deal state operations
 ////////////////////////////////////////////////////////////////////////////////
 
-func (st *State) updatePendingDealStatesForParty(rt Runtime, addr addr.Address) (amountSlashedTotal abi.TokenAmount) {
-	// For consistency with HandleExpiredDeals, only process updates up to the end of the _previous_ epoch.
-	epoch := rt.CurrEpoch() - 1
-
-	dbp, err := AsSetMultimap(adt.AsStore(rt), st.DealIDsByParty)
-	if err != nil {
-		rt.Abortf(exitcode.ErrIllegalState, "failed to load dead ids set: %s", err)
-	}
-
-	var extractedDealIDs []abi.DealID
-	err = dbp.ForEach(addr, func(id abi.DealID) error {
-		extractedDealIDs = append(extractedDealIDs, id)
-		return nil
-	})
-	if err != nil {
-		rt.Abortf(exitcode.ErrIllegalState, "foreach error %v", err)
-	}
+func (st *State) updatePendingDealState(rt Runtime, dealID abi.DealID, epoch abi.ChainEpoch) (abi.TokenAmount, abi.ChainEpoch) {
+	amountSlashed := abi.NewTokenAmount(0)
 
-	amountSlashedTotal = st.updatePendingDealStates(rt, extractedDealIDs, epoch)
-	return
-}
-
-func (st *State) updatePendingDealStates(rt Runtime, dealIDs []abi.DealID, epoch abi.ChainEpoch) abi.TokenAmount {
-	amountSlashedTotal := abi.NewTokenAmount(0)
-
-	for _, dealID := range dealIDs {
-		amountSlashedTotal = big.Add(amountSlashedTotal, st.updatePendingDealState(rt, dealID, epoch))
-	}
-
-	return amountSlashedTotal
-}
-
-// TODO: This does waaaay too many redundant hamt reads
-func (st *State) updatePendingDealState(rt Runtime, dealID abi.DealID, epoch abi.ChainEpoch) (amountSlashed abi.TokenAmount) {
-	amountSlashed = abi.NewTokenAmount(0)
-
-	deal := st.mustGetDeal(rt, dealID)
 	state := st.mustGetDealState(rt, dealID)
 
 	everUpdated := state.LastUpdatedEpoch != epochUndefined
 	everSlashed := state.SlashEpoch != epochUndefined
 
 	Assert(!everUpdated || (state.LastUpdatedEpoch <= epoch)) // if the deal was ever updated, make sure it didn't happen in the future
 
-	if state.LastUpdatedEpoch == epoch { // TODO: This looks fishy, check all places that set LastUpdatedEpoch
-		return
-	}
+	deal := st.mustGetDeal(rt, dealID)
 
 	if state.SectorStartEpoch == epochUndefined {
 		// Not yet appeared in proven sector; check for timeout.
 		if epoch > deal.StartEpoch {
-			return st.processDealInitTimedOut(rt, dealID)
+			return st.processDealInitTimedOut(rt, dealID, deal, state), 0
 		}
-		return
+		return amountSlashed, 0
```

Comment:
I actually think the case youre commenting on will never be hit, it would mean that we called this method for a deal that hasnt been proven, before its starting deadline. Nothing registers a callback for deals at that point, so we really shouldnt ever hit that. I could call abort here, but i'm not 100% sure